shared-libs/transitions/src/transitions/create_user_for_contacts.js

```diff
@@ -148,36 +153,68 @@ module.exports = {
       );
     }
   },
-  filter: (doc) => {
+  filter: (doc, info = {}) => {
     const contactType = contactTypeUtils.getContactType(config.getAll(), doc);
     if (
       !contactType
       || !contactTypeUtils.isPersonType(contactType)
       || !doc.user_for_contact
-      || !doc.user_for_contact.replace
     ) {
       return false;
     }
 
-    return !!Object
-      .values(doc.user_for_contact.replace)
-      .find(({ status }) => status === USER_CREATION_STATUS.READY);
+    const isReplacingUser = !!doc.user_for_contact.replace;
+    const isAddingUser = doc.user_for_contact.create && doc.user_for_contact.create.add === 'true';
+    const hasAlreadyCreatedUser = !!(info && info.transitions && info.transitions.create_user_for_contacts);
+    if (isReplacingUser && isAddingUser) {
+      const hasFinishedReplacing = Object
+        .values(doc.user_for_contact.replace)
+        .every(({ status }) => [USER_CREATION_STATUS.COMPLETE, USER_CREATION_STATUS.ERROR].includes(status));
+
+      return hasFinishedReplacing && isAddingUser && !hasAlreadyCreatedUser;
+    }
+
+    if (isReplacingUser) {
+      return Object
+        .values(doc.user_for_contact.replace)
+        .some(({ status }) => status === USER_CREATION_STATUS.READY);
+    }
+
+    if (isAddingUser) {
+      return !hasAlreadyCreatedUser;
+    }
+
+    return false;
   },
   onMatch: async change => {
-    const usersToReplace = getUsersToReplace(change.doc);
-    const replaceUsers = usersToReplace.map(user => replaceUser(change.doc, user));
-    const replacementResults = await Promise.allSettled(replaceUsers);
-    const errors = replacementResults
-      .filter(({ status }) => status === 'rejected')
-      .map(({ reason }) => reason);
-    if (errors.length) {
+    const isReplacingUser = Boolean(change.doc.user_for_contact.replace);
+    if (isReplacingUser) {
```

Comment:
I've been mulling this over and I don't think we should run those two actions in parallel. We could run into a race condition where the user wasn't created yet and the replace fails to process because of the not-yet-existing user

Reply:
I'm coming up with something that keeps the syntax as nice and readable as your suggestion but runs each action sequentially

Reply:
> We could run into a race condition where the user wasn't created yet and the replace fails to process because of the not-yet-existing user

It if fine with me to not run these processes in parallel, but I did want to note that I don't think we have to worry about this scenario where we need to replace the newly created user.  This is because the data structure in `user_for_contact.replace` requires the actual _username_ of the user being replaced (which will not exist for the new user before we randomly generate it when creating the user).  I think this means it is not possible to create and replace the _same user_ at the _same time_.